Update connection management documentation (#1834)
nest/nest-simulator: doc/guides/connection_management.rst

Connection Management
=====================

In NEST, connections are created with the `Connect()` function. You can
create connections with different types of connectivity patterns, which

> are defined inside the function under the ``conn_spec`` argument, and the

Reviewer: ```suggestion
are defined inside the function under the ``conn_spec`` argument and the
```
```suggestion
are defined inside the function under the ``conn_spec`` argument, and the
```

> key 'rule'. The patterns available are described in
> :ref:`Connection Rules <conn_rules>`. In addition the synapse model can

Reviewer: ```suggestion
key 'rule'. The patterns available are described in the
:ref:`Connection Rules <conn_rules>` section. In addition, the synapse model can
```
```suggestion
key 'rule'. The patterns available are described in
:ref:`Connection Rules <conn_rules>`. In addition the synapse model can
```

> be specified within the Connect function and all synaptic parameters can

Reviewer: ```suggestion
be specified within the `Connect()` function, and all synaptic parameters can
```

be randomly distributed. More information about synapse models and their

> parameters can be found in :ref:`Synapse Specification <synapse_spec>`.

Reviewer: ```suggestion
parameters can be found in the :ref:`Synapse Specification <synapse_spec>` section.
```

> In addition switches allowing self-connections (``allow_autapses``, default:

Reviewer: ```suggestion
In addition, switches allowing self-connections (``allow_autapses``, default:
```

True) and multiple connections between pairs of neurons (``allow_multapses``,
default: True) can be contained in the dictionary. The validity of the
switches is confined by the Connect-call. Thus, by connecting the same set
of neurons multiple times with the switch 'allow_multapses' set to False, one
particular connection might still be established multiple times.

``syn_spec`` defines the synapse type and its properties. It can be
given as a string defining the synapse model (default:
'static_synapse'), as an object defining :ref:`collocated synapses <collocated_synapses>`,

> or as a dictionary. By using the key-word variant

Reviewer: ```suggestion
or as a dictionary. By using the keyword variant
```

> (``Connect(pre, post, syn_spec=syn_spec_dict)``), the ``conn_spec`` can be

Reviewer: ```suggestion
(``Connect(pre, post, syn_spec=syn_spec_dict)``), ``conn_spec`` can be
```

default. The exact usage of the synapse dictionary is described in
:ref:`synapse_spec`.

> After a connection is established it might be beneficial to look up the number of

Reviewer: ``` suggestion
After a connection is established, it might be beneficial to look up the number of
```

and its parameters, such as in- or out-degrees, is required.

one-to-one
~~~~~~~~~~

.. image:: ../_static/img/One_to_one.png
     :width: 200px
     :align: center

The ith node in ``pre`` is connected to the ith node in ``post``. The

> NodeCollections of pre and post have to be of the same length.

Reviewer: ```suggestion
NodeCollections of ``pre`` and ``post`` have to be of the same length.
```

and its parameters, such as in- or out-degrees, is required.

one-to-one
~~~~~~~~~~

.. image:: ../_static/img/One_to_one.png
     :width: 200px
     :align: center

The ith node in ``pre`` is connected to the ith node in ``post``. The
NodeCollections of ``pre`` and ``post`` have to be of the same length.

::

    n = 10
    A = nest.Create('iaf_psc_alpha', n)
    B = nest.Create('spike_recorder', n)
    nest.Connect(A, B, 'one_to_one')

> all-to-all

Author: Done in [1b80b22](https://github.com/nest/nest-simulator/pull/1834/commits/1b80b22bc0996a9570cbea9c58a76e2794582ff2).

::

    n, m, p = 10, 12, 0.2 
    A = nest.Create('iaf_psc_alpha', n) 
    B = nest.Create('iaf_psc_alpha', m) 
    conn_dict = {'rule': 'symmetric_pairwise_bernoulli', 'p': p, 'allow_autapses': False, 'make_symmetric': True} 
    nest.Connect(A, B, conn_dict)

.. _synapse_spec:

Synapse Specification
---------------------

The synapse properties can be given as a string, a ``CollocatedSynapse``
object, or a dictionary. The string can be the name of a pre-defined
synapse which can be found in the synapsedict (see  :ref:`synapse-types`)
or a manually defined synapse via `CopyModel()`.

::

    n = 10
    A = nest.Create('iaf_psc_alpha', n)
    B = nest.Create('iaf_psc_alpha', n)
    nest.CopyModel('static_synapse','excitatory',{'weight': 2.5, 'delay': 0.5})
    nest.Connect(A, B, syn_spec='excitatory')

Specifying the synapse properties in a dictionary allows for distributed
synaptic parameters. In addition to the key ``synapse_model``, the dictionary can
contain specifications for ``weight``, ``delay``, ``receptor_type`` and
parameters specific to the chosen synapse model. The specification of
all parameters is optional. Unspecified parameters will use the default

> values determined by the current synapse model. All parameters can be

Author: Done in [0239897](https://github.com/nest/nest-simulator/pull/1834/commits/02398970ea954faf592d2c23952cf7c61048f505).

scalars, arrays or distributions (specified as a ``nest.Parameter``). One
synapse dictionary can contain an arbitrary combination of parameter
types, as long as they agree with the connection routine (``rule``).

Scalar parameters
~~~~~~~~~~~~~~~~~

Scalar parameters must be given as floats except for the
'receptor_type' which has to be initialized as an integer. For more
information on the receptor type see :ref:`receptor-types`. When a synapse
parameter is given as a scalar, the value will be applied to all connections. 

::

    n = 10
    neuron_dict = {'tau_syn': [0.3, 1.5]}
    A = nest.Create('iaf_psc_exp_multisynapse', n, neuron_dict)
    B = nest.Create('iaf_psc_exp_multisynapse', n, neuron_dict)
    syn_dict ={'synapse_model': 'static_synapse', 'weight': 2.5, 'delay': 0.5, 'receptor_type': 1}
    nest.Connect(A, B, syn_spec=syn_dict)

Array parameters
~~~~~~~~~~~~~~~~

Array parameters can be used in conjunction with the rules
``one_to_one``, ``all_to_all``, ``fixed_indegree``, ``fixed_outdegree``
and ``fixed_total_number``. The arrays can be specified as numpy arrays or
lists. As with the scalar parameters, all parameters but the receptor

> types must be specified as arrays of floats. For ``one_to_one`` the

Author: Done in [1b80b22](https://github.com/nest/nest-simulator/pull/1834/commits/1b80b22bc0996a9570cbea9c58a76e2794582ff2).

is created with probability ``p`` from ``pre`` to ``post``, as well as
a connection from ``post`` to ``pre`` (two connections in total). To
use the 'symmetric_pairwise_bernoulli' rule, ``allow_autapses`` must
be False, and ``make_symmetric`` must be True.

Example:

::

    n, m, p = 10, 12, 0.2 
    A = nest.Create('iaf_psc_alpha', n) 
    B = nest.Create('iaf_psc_alpha', m) 
    conn_dict = {'rule': 'symmetric_pairwise_bernoulli', 'p': p, 'allow_autapses': False, 'make_symmetric': True} 
    nest.Connect(A, B, conn_dict)

.. _synapse_spec:

Synapse Specification
---------------------

The synapse properties can be given as a string, a ``CollocatedSynapse``
object, or a dictionary. The string can be the name of a pre-defined
synapse which can be found in the synapsedict (see  :ref:`synapse-types`)
or a manually defined synapse via `CopyModel()`.

Example:

::

    n = 10
    A = nest.Create('iaf_psc_alpha', n)
    B = nest.Create('iaf_psc_alpha', n)
    nest.CopyModel('static_synapse','excitatory',{'weight': 2.5, 'delay': 0.5})
    nest.Connect(A, B, syn_spec='excitatory')

Specifying the synapse properties in a dictionary allows for distributed

> synaptic parameters. In addition to the key ``synapse_model`` the dictionary can

Reviewer: ```suggestion
synaptic parameters. In addition to the key ``synapse_model``, the dictionary can
```

contain specifications for ``weight``, ``delay``, ``receptor_type`` and
parameters specific to the chosen synapse model. The specification of
all parameters is optional. Unspecified parameters will use the default
values determined by the current synapse model. All parameters can be

> scalars, arrays or distributions (specified as a ``nest.Parameter``). One

Reviewer: Perhaps we could have parameter subsections titled "scalar", "array" and "distributed". What do you think?

Author: Will do!

or list with shape `(len(post), indegree)`, where indegree is the number of
incoming connections per target neuron. Therefore, the rows describe the
target and the columns the connections converging to the target neuron,
regardless of the identity of the source neurons.

Example:

::

    A = nest.Create('iaf_psc_alpha', 5)
    B = nest.Create('iaf_psc_alpha', 3)
    conn_dict = {'rule': 'fixed_indegree', 'indegree': 2}
    syn_dict = {'weight': [[1.2, -3.5],[0.4, -0.2],[0.6, 2.2]]}
    nest.Connect(A, B, conn_spec=conn_dict, syn_spec=syn_dict)

For ``fixed_outdegree`` the array has to be a two-dimensional NumPy array
or list with shape `(len(pre), outdegree)`, where outdegree is the number of
outgoing connections per source neuron. Therefore, the rows describe the
source and the columns the connections starting from the source neuron
regardless of the identity of the target neuron.

Example:

::

    A = nest.Create('iaf_psc_alpha', 2)
    B = nest.Create('iaf_psc_alpha', 5)
    conn_dict = {'rule': 'fixed_outdegree', 'outdegree': 3}
    syn_dict = {'weight': [[1.2, -3.5, 0.4], [-0.2, 0.6, 2.2]]}
    nest.Connect(A, B, conn_spec=conn_dict, syn_spec=syn_dict)

> For ``fixed_total_number`` the array has to be same the length as the

Reviewer: ```suggestion
For ``fixed_total_number``, the array has to be same the length as the
```

    syn_spec = nest.CollocatedSynapses({'weight': 4., 'delay': 1.5},
                                       {'synapse_model': 'stdp_synapse'},
                                       {'synapse_model': 'stdp_synapse', 'alpha': 3.})
    nest.Connect(nodes, nodes, conn_spec='one_to_one', syn_spec=syn_spec)

    conns = nest.GetConnections()
    print(conns.alpha)

This will create 9 connections: 3 using `static_synapse` with a `weight` of `4.0` and `delay` of `1.5`, and 6 using
the `stdp_synapse`. Of the 6 using `stdp_synapse`, 3 will have the default alpha value, and 3 will have an alpha of
`3.0`.

  >>> print(nest.GetKernelStatus('num_connections'))
  9

If you want to connect with different :ref:`receptor types <receptor-types>`, you can do the following:

  ::

    src = nest.Create('iaf_psc_exp_multisynapse', 7)
    trgt = nest.Create('iaf_psc_exp_multisynapse', 7, {'tau_syn': [0.1 + i for i in range(7)]})

    syn_spec = nest.CollocatedSynapses({'weight': 5.0, 'receptor_type': 2},
                                       {'weight': 1.5, 'receptor_type': 7})

    nest.Connect(src, trgt, 'one_to_one', syn_spec=syn_spec)

    conns = nest.GetConnections()
    print(conns.get())

> You can see how many synapse parameters you have by doing `len()` on your `CollocatedSynapses` object:

Reviewer: ```suggestion
You can see how many synapse parameters you have by calling `len()` on your `CollocatedSynapses` object:
```

  >>> len(syn_spec)
  2

.. _dist_params:

Distributed parameters
~~~~~~~~~~~~~~~~~~~~~~

**Distributed** parameters are given as ``nest.Parameter`` objects that represent
values drawn from random distributions. These distributions can either be based on
spatial node parameters, on default values, or on constant distribution values
you provide. It is possible to combine the parameters to create distributions
tailor made for your needs.

The following parameters and functionalities are provided:

- Random parameters
- Spatial parameters
- Spatially distributed parameters
- Mathematical functions
- Clipping, redrawing, and conditional parameters

For more information, check out the documentation on the different

> :doc:`PyNEST api's <../ref_material/pynest_apis>`, or the section on parametrization

Reviewer: ```suggestion
:doc:`PyNEST APIs <../ref_material/pynest_apis>`, or the section on parametrization
```

> in our :doc:`nest2_to_nest3/nest2_to_nest3_overview` guide.

Reviewer: Should we add a ``:ref:`` to the section?

Author: Yes! I wanted to do that, I just didn't know how, I will update :)

    A4 = nest.Create('iaf_psc_alpha')
    B = nest.Create('iaf_cond_alpha_mc')

    receptors = nest.GetDefaults('iaf_cond_alpha_mc')['receptor_types']
    print(receptors)
        {'soma_exc': 1,
         'soma_inh': 2,
         'soma_curr': 7,
         'proximal_exc': 3
         'proximal_inh': 4,
         'proximal_curr': 8,
         'distal_exc': 5,
         'distal_inh': 6,
         'distal_curr': 9,}

    nest.Connect(A1, B, syn_spec={'receptor_type': receptors['distal_inh']})
    nest.Connect(A2, B, syn_spec={'receptor_type': receptors['proximal_inh']})
    nest.Connect(A3, B, syn_spec={'receptor_type': receptors['proximal_exc']})
    nest.Connect(A4, B, syn_spec={'receptor_type': receptors['soma_inh']})

.. _synapse-types:

Synapse Types
-------------

NEST supports multiple synapse types that can be specified during
connection setup. The default synapse type in NEST is the
``static_synapse``. Its weight does not change over time. To allow
learning and plasticity, it is possible to use other synapse types that
implement long-term or short-term plasticity. A list of available types
is accessible via the command ``nest.Models('synapses')``. The output of this

> command (as of commit ``1159664``) is shown below:

Reviewer: Should we link the commit?

connections. They can be used to save memory.

The default values of a synapse type can be inspected using the command
``nest.GetDefaults()``, which takes the name of the synapse as an argument,
and modified with ``nest.SetDefaults()``, which takes the name of the synapse
type and a parameter dictionary as arguments.

::

    print(nest.GetDefaults('static_synapse'))

        {'delay': 1.0,
         'has_delay': True,
         'num_connections': 0,
         'receptor_type': 0,
         'requires_symmetric': False,
         'sizeof': 32,
         'synapse_model':
         'static_synapse',
         'weight': 1.0,
         'weight_recorder': ()}

    nest.SetDefaults('static_synapse', {'weight': 2.5})

For the creation of custom synapse types from already existing synapse
types, the command ``nest.CopyModel`` is used. It has an optional argument
``params`` to directly customize it during the copy operation. Otherwise
the defaults of the copied model are taken.

::

    nest.CopyModel('static_synapse', 'inhibitory', {'weight': -2.5})
    nest.Connect(A, B, syn_spec='inhibitory')

> **Note**: Not all nodes can be connected via all available synapse

Reviewer: We could turn this into an RST note.

         'target_thread': [0, 0],
         'weight': [1.0, 1.0]}

> The ``get()`` function also takes a string or list of strings as argument. You

Reviewer: ``` suggestion
The ``get()`` function also takes a string or list of strings as arguments. You
```

can thus retrieve specific parameters if you do not want to inspect the entire
synapse dictionary:

  >>>  conn.get('weight')
       [1.0, 1.0]

  >>>  conn.get(['source', 'target'])
       {'source': [1, 2], 'target': [3, 3]}

Another way of retrieving specific parameters is by getting it directly from
the SynapseCollection:

    >>>  conn.delay
         [1.0, 1.0]

For :doc:`spatially distributed networks <spatial/index>`, you can access the distance between
the source-target pairs by calling `distance` on your SynapseCollection.

>>>  spatial_conn.distance
     (0.47140452079103173,
      0.33333333333333337,
      0.4714045207910317,
      0.33333333333333337,
      3.925231146709438e-17,
      0.33333333333333326,
      0.4714045207910317,
      0.33333333333333326,
      0.47140452079103157)

You can further examine the SynapseCollection by checking the length of the object.

> Or, by printing it, which will return a table of source and target node IDs:

Reviewer: ``` suggestion
You can further examine the SynapseCollection by checking the length of the object
or by printing it, which will return a table of source and target node IDs:
```

         'target': 2,
         'target_thread': 0,
         'weight': 2.0}

Updating a single parameter is done by calling ``set(parameter_name=parameter_value)``.
You can use a single value, a list, or a ``nest.Parameter`` as value. If a single

> value is given, the value is set on all connections, while if you use a list to set

Reviewer: ``` suggestion
value is given, the value is set on all connections. If you use a list to set
```

the parameter, the list needs to be the same length as the SynapseCollection.

  >>>  conn.set(weight=[4.0, 4.5, 5.0, 5.5])

Just as you can retrieve several parameters at once with the ``get()`` function

> above, you can also set several parameters at once, with

Reviewer: ``` suggestion
above, you can also set several parameters at once with
```

> ``set(parameter_dictionary)``. You can again use a single value, a list, or a

Reviewer: ``` suggestion
``set(parameter_dictionary)``. Again, you can again use a single value, a list, or a
```

``nest.Parameter`` as value. 

  >>>  conn.set({'weight': [1.5, 2.0, 2.5, 3.0], 'delay': 2.0})

You can also directly set parameters of your SynapseCollection:

  >>>  conn.weight = 5.
  >>>  conn.weight
       [5.0, 5.0, 5.0, 5.0]
  >>>  conn.delay = [5.1, 5.2, 5.3, 5.4]
  >>>  conn.delay
       [5.1, 5.2, 5.3, 5.4]

> Note that some parameters, like `source` and `target`, cannot be set.  The documentation of a specific

Reviewer: ``` suggestion
Note that some parameters, like `source` and `target`, cannot be set. The documentation of a specific
```
